### PR TITLE
Reduce allocations in HandleCompilationStart

### LINKED ISSUE #1687: Finish reducing allocations in HandleCompilationStart
#1683 partially reduced allocations (documentation in layout rules). The remaining rules need to be updated to match.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #1687.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1300ElementMustBeginWithUpperCaseLetter.cs [hunk 1]
```diff
@@ -60,18 +69,18 @@ private static void HandleCompilationStart(CompilationStartAnalysisContext conte
...
-            context.RegisterSyntaxNodeActionHonorExclusions(HandleEnumDeclarationSyntax, SyntaxKind.EnumDeclaration);
-            context.RegisterSyntaxNodeActionHonorExclusions(HandleStructDeclarationSyntax, SyntaxKind.StructDeclaration);
-            context.RegisterSyntaxNodeActionHonorExclusions(HandleDelegateDeclarationSyntax, SyntaxKind.DelegateDeclaration);
-            context.RegisterSyntaxNodeActionHonorExclusions(HandleEventDeclarationSyntax, SyntaxKind.EventDeclaration);
-            context.RegisterSyntaxNodeActionHonorExclusions(HandleEventFieldDeclarationSyntax, SyntaxKind.EventFieldDeclaration);
-            context.RegisterSyntaxNodeActionHonorExclusions(HandleMethodDeclarationSyntax, SyntaxKind.MethodDeclaration);
-            context.RegisterSyntaxNodeActionHonorExclusions(HandlePropertyDeclarationSyntax, SyntaxKind.PropertyDeclaration);
+            context.RegisterSyntaxNodeActionHonorExclusions(NamespaceDeclarationAction, SyntaxKind.NamespaceDeclaration);
+            context.RegisterSyntaxNodeActionHonorExclusions(ClassDeclarationAction, SyntaxKind.ClassDeclaration);
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1300ElementMustBeginWithUpperCaseLetter.cs]: I think we should have global cache for common `SyntaxKind`s like this one.

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1300ElementMustBeginWithUpperCaseLetter.cs]: You mean [something like this](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/AnalyzerExtensions.cs#L145-L162)?

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1300ElementMustBeginWithUpperCaseLetter.cs]: Nice that we already have it)

**Comment by Maxwe11** [comment]
LGTM

File: StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1300ElementMustBeginWithUpperCaseLetter.cs (hunk 1 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1300ElementMustBeginWithUpperCaseLetter.cs]: Yep, and I removed the `params` overload of `RegisterSyntaxNodeActionHonorExclusions` to help ensure these allocations do not recur in future analyzer implementations. :smile:

**Pull request merged**